ksql-cli/src/test/java/io/confluent/ksql/cli/SslFunctionalTest.java

```diff
@@ -0,0 +1,233 @@
+    // Then:
+    expectedException.expect(SSLHandshakeException.class);
+    expectedException.expectCause(hasCause(hasMessage(containsString(
+        "unable to find valid certification path to requested target"))));
+
+    // When:
+    makeWsRequest(false);
+  }
+
+  @Test
+  public void shouldBeAbleToUseWss() throws Exception {
+    // When:
+    final Code result = makeWsRequest(true);
+
+    // Then:
+    assertThat(result, is(Code.OK));
+  }
+
+  private void givenTrustStoreConfigured() {
+    clientProps = TRUST_STORE_PROPS;
+  }
+
+  private Code canMakeCliRequest() {
+    final String serverAddress = REST_APP.getHttpsListener().toString();
+
+    try (KsqlRestClient restClient = new KsqlRestClient(serverAddress, emptyMap(), clientProps)) {
+
+      final RestResponse<?> response = restClient.makeKsqlRequest("show topics;");
+      if (response.isSuccessful()) {
+        return Code.OK;
+      }
+
+      return HttpStatus.getCode(response.getErrorMessage().getErrorCode());
+    }
+  }
+
+  private static Code makeWsRequest(final boolean setUpTrustStore) throws Exception {
+    final SslContextFactory sslContextFactory = new SslContextFactory();
+    if (setUpTrustStore) {
+      sslContextFactory.setTrustStorePath(ClientTrustStore.trustStorePath());
+      sslContextFactory.setTrustStorePassword(ClientTrustStore.trustStorePassword());
+      sslContextFactory.setEndpointIdentificationAlgorithm("");
+    }
+
+    final HttpClient httpClient = new HttpClient(sslContextFactory);
+    httpClient.start();
+    final WebSocketClient wsClient = new WebSocketClient(httpClient);
+    wsClient.start();
+
+    try {
+      final ClientUpgradeRequest request = new ClientUpgradeRequest();
+      final WebSocketListener listener = new WebSocketListener();
+      final URI wsUri = REST_APP.getWssListener().resolve("/ws/query");
+
+      wsClient.connect(listener, wsUri, request);
+
+      assertThat("Response received",
+          listener.latch.await(30, TimeUnit.SECONDS), is(true));
+
+      final Throwable error = listener.error.get();
+      return error == null ? Code.OK : extractStatusCode(error);
+    } finally {
+      wsClient.stop();
+      httpClient.destroy();
+    }
+  }
+
+  private static Code extractStatusCode(final Throwable t) throws Exception {
+    if (t instanceof UpgradeException) {
+      final Code code = HttpStatus.getCode(((UpgradeException) t).getResponseStatusCode());
+      if (code == null) {
+        throw new RuntimeException(t.getCause());
+      }
+      return code;
+    }
+
+    if (t instanceof Exception) {
+      throw (Exception) t;
+    }
+
+    throw new RuntimeException(t);
+  }
+
+  @SuppressWarnings("unused")
+  @WebSocket
+  public static class WebSocketListener {
+
+    final CountDownLatch latch = new CountDownLatch(1);
+    final AtomicReference<Throwable> error = new AtomicReference<>();
+
+    @OnWebSocketConnect
+    public void onConnect(final Session session) {
+      session.close();
```

Comment:
i recall seeing a test where this pattern was flaky for some reason - can you run this 100x times to make sure it's not flakey?

Reply:
Sure.